nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java

```diff
@@ -90,4 +113,148 @@ private static void invalidInstantiationError(
+      // annotations preserved
+      rhsType =
+          typeWithPreservedAnnotations(
+              (ParameterizedTypeTree) ((NewClassTree) rhsTree).getIdentifier());
+    }
+    if (lhsType != null && rhsType != null) {
+      compareAnnotations((Type.ClassType) lhsType, (Type.ClassType) rhsType, tree);
+    }
+  }
+
+  private void compareAnnotations(Type.ClassType lhsType, Type.ClassType rhsType, Tree tree) {
+    Types types = state.getTypes();
+    rhsType = (Type.ClassType) types.asSuper(rhsType, lhsType.tsym);
+    if (rhsType == null) {
+      throw new RuntimeException("did not find supertype of " + rhsType + " matching " + lhsType);
+    }
+    List<Type> lhsTypeArguments = lhsType.getTypeArguments();
+    List<Type> rhsTypeArguments = rhsType.getTypeArguments();
+    if (lhsTypeArguments.size() != rhsTypeArguments.size()) {
+      throw new RuntimeException(rhsType + " does not match " + lhsType);
+    }
+    for (int i = 0; i < lhsTypeArguments.size(); i++) {
+      com.sun.tools.javac.util.List<Attribute.TypeCompound> annotationMirrorsLHS =
+          lhsTypeArguments.get(i).getAnnotationMirrors();
+      com.sun.tools.javac.util.List<Attribute.TypeCompound> annotationMirrorsRHS =
+          rhsTypeArguments.get(i).getAnnotationMirrors();
+      boolean isLHSNullableAnnotated =
+          Nullness.hasNullableAnnotation(annotationMirrorsLHS.stream(), config);
+      boolean isRHSNullableAnnotated =
+          Nullness.hasNullableAnnotation(annotationMirrorsRHS.stream(), config);
+      if (isLHSNullableAnnotated != isRHSNullableAnnotated) {
+        invalidAssignmentInstantiationError(tree, lhsType, rhsType, state, analysis);
+        return;
+      }
+      // nested generics
+      if (lhsTypeArguments.get(i).getTypeArguments().length() > 0) {
+        compareAnnotations(
+            (Type.ClassType) lhsTypeArguments.get(i),
+            (Type.ClassType) rhsTypeArguments.get(i),
+            tree);
+      }
+    }
+  }
+
+  /**
+   * @param tree A parameterized typed tree for which we need class type with preserved annotations.
+   * @return A Type with preserved annotations.
+   */
+  private Type.ClassType typeWithPreservedAnnotations(ParameterizedTypeTree tree) {
+    Type.ClassType type = (Type.ClassType) ASTHelpers.getType(tree);
+    Preconditions.checkNotNull(type);
+    Type nullableType = NULLABLE_TYPE_SUPPLIER.get(state);
+    List<? extends Tree> typeArguments = tree.getTypeArguments();
+    List<Type> newTypeArgs = new ArrayList<>();
+    for (int i = 0; i < typeArguments.size(); i++) {
+      List<Attribute.TypeCompound> allAnnotations = new ArrayList<>();
+      List<JCTree.JCAnnotation> annotations = new ArrayList<>();
+      if (typeArguments.get(i).getClass().equals(JCTree.JCAnnotatedType.class)) {
+        JCTree.JCAnnotatedType annotatedType = (JCTree.JCAnnotatedType) typeArguments.get(i);
+        annotations = annotatedType.getAnnotations();
```

Comment:
Just assign the `annotatedType` variable in the `if` condition, and then declare this variable and assign it after the `if`, removing the duplication with line 229